File: lib/src/main/java/com/dhy/retrofitrxutil/ObserverX.java
package com.dhy.retrofitrxutil;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public abstract class ObserverX<T> implements Observer<T>, IObserverX {
    private static IErrorHandler defaultErrorHandler;
    private static StyledProgressGenerator defaultStyledProgressGenerator;

    public static void setDefaultErrorHandler(IErrorHandler defaultErrorHandler) {
        ObserverX.defaultErrorHandler = defaultErrorHandler;
    }

    public static void setDefaultStyledProgressGenerator(StyledProgressGenerator defaultStyledProgressGenerator) {
        ObserverX.defaultStyledProgressGenerator = defaultStyledProgressGenerator;
    }

    @NonNull
    private Context context;
    protected StyledProgress styledProgress;
    protected final boolean successOnly, autoDismiss;
    private Disposable disposable;

    public ObserverX(@NonNull Context context) {
        this(context, true, true);
    }

    /**
     * @param autoDismiss autoDismissProgress
     */
    public ObserverX(@NonNull Context context, boolean successOnly, boolean autoDismiss) {
        this.context = context;
        this.successOnly = successOnly;
        this.autoDismiss = autoDismiss;
        styledProgress = getStyledProgress();
        if (styledProgress == null) styledProgress = StyledProgressOfNone.getInstance();
    }

    @NonNull
    public Context getContext() {
        return context;
    }

    /**
     * Override this for custom progress
     *
     * @return null for StyledProgressOfNone
     */
    @Nullable
    protected StyledProgress getStyledProgress() {
        return defaultStyledProgressGenerator.generate(this);
    }

    @Override
    public void onSubscribe(Disposable disposable) {
        this.disposable = disposable;
        styledProgress.showProgress();
        if (context instanceof IDisposable) {
            ((IDisposable) context).registerDisposable(context, disposable);
        }
    }

    @Override
    public void onNext(T t) {
        if (successOnly && t instanceof IResponseStatus) {
            IResponseStatus status = (IResponseStatus) t;
            if (status.isSuccess()) {
                onResponse(t);
            } else {
                onFailed(status);
            }
        } else onResponse(t);
    }

    protected void onFailed(@NonNull IResponseStatus status) {
        onError(new ThrowableBZ(status));
    }

    /**
     * Override this for custom ErrorHandler
     */
    @NonNull
    protected IErrorHandler getErrorHandler() {
        return defaultErrorHandler;
    }

    @Override
    public void onError(Throwable e) {
        getErrorHandler().onError(this, e);
    }

    @Override
    public void onComplete() {
        cancel();
        if (autoDismiss) dismissProgress();
    }

    public void showProgress() {
        styledProgress.showProgress();
    }

    public void dismissProgress() {
        styledProgress.dismissProgress();
    }

    protected abstract void onResponse(T response);

    public void cancel() {
        if (disposable != null) {
            disposable.dispose();
            if (context instanceof IDisposable) {
                ((IDisposable) context).onComplete(context, disposable);
            }
        }
    }
}

File: lib/src/main/java/com/dhy/retrofitrxutil/ObserverX.kt
package com.dhy.retrofitrxutil

import android.app.Activity
import android.content.Context
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable

abstract class ObserverX<T>(final override val context: Context, private val successOnly: Boolean = true) : Observer<T>, IObserverX {
    private var disposable: Disposable? = null
    private val progress by lazy { getStyledProgress() }

    /**
     * Override this for custom progress
     *
     * @return null for StyledProgressOfNone
     */
    protected open fun getStyledProgress(): StyledProgress? {
        return defaultStyledProgressGenerator?.generate(this)
    }

    override fun onSubscribe(disposable: Disposable) {
        this.disposable = disposable
        showProgress()
        if (context is Activity) requests[this] = context
    }

    internal fun onDestroy() {
        cancel()
    }

    private fun removeObserver() {
        if (context is Activity) requests.remove(this)
    }

    override fun onNext(t: T) {
        dismissProgress(true)
        if (successOnly && t is IResponseStatus) {
            val status = t as IResponseStatus
            if (status.isSuccess) {
                onResponse(t)
            } else {
                onFailed(status)
            }
        } else onResponse(t)
    }

    protected open fun onFailed(status: IResponseStatus) {
        onError(ThrowableBZ(status))
    }

    /**
     * Override this for custom ErrorHandler
     */
    protected open val errorHandler: IErrorHandler
        get() = defaultErrorHandler!!

    final override fun onError(e: Throwable) {
        dismissProgress(false)
        onErrorResponse(e)
    }

    open fun onErrorResponse(e: Throwable) {
        errorHandler.onError(this, e)
    }

    override fun onComplete() {
        disposable = null
    }

    private fun showProgress() {
        if (!isFinishing()) progress?.showProgress()
    }

    private fun isFinishing(): Boolean {
        return context is Activity && context.isFinishing
    }

    override fun dismissProgress(delay: Boolean) {
        progress?.dismissProgress(delay)
        removeObserver()
    }

    protected abstract fun onResponse(response: T)
    override fun cancel() {
        disposable?.dispose()
        dismissProgress(false)
    }

    override fun isCanceled(): Boolean {
        return disposable == null || disposable?.isDisposed == true
    }

    companion object {
        private val requests: MutableMap<ObserverX<*>, Activity> = mutableMapOf()
        internal fun onActivityDestroyed(activity: Activity) {
            ArrayList(requests.keys).forEach {
                val v = requests[it]
                if (v == activity) it.onDestroy()
            }
        }

        private var defaultErrorHandler: IErrorHandler? = null
        private var defaultStyledProgressGenerator: StyledProgressGenerator? = null

        @JvmStatic
        @Deprecated("use setErrorHandler", replaceWith = ReplaceWith("ObserverX.setErrorHandler(handler)"))
        fun setDefaultErrorHandler(handler: IErrorHandler?) {
            defaultErrorHandler = handler
        }

        @JvmStatic
        fun setErrorHandler(handler: IErrorHandler?) {
            defaultErrorHandler = handler
        }

        @JvmStatic
        @Deprecated("use setProgressGenerator", replaceWith = ReplaceWith("ObserverX.setProgressGenerator(generator)"))
        fun setDefaultStyledProgressGenerator(generator: StyledProgressGenerator?) {
            defaultStyledProgressGenerator = generator
        }

        @JvmStatic
        fun setProgressGenerator(generator: StyledProgressGenerator?) {
            defaultStyledProgressGenerator = generator
        }
    }
}